src/utils/gadm_gpkg_processing.py

```diff
@@ -0,0 +1,200 @@
+
+
+##
+#
+def is_int(string):
+    try:
+        int(string)
+        return True
+    except ValueError:
+        return False
+
+##
+# Aggregate lower admin levels for specified country
+def get_country_admin_levels(gdf_world, country_aoi):
+    print('Checking {0} Admin Levels.'.format(country_aoi))
+    return_dict = {}
+    gdf = gdf_world.loc[gdf_world['GID_0'] == country_aoi].copy()
+    # Determine level of admin - check GID_* cols for None values.
+    cols = gdf.columns.values.tolist()
+    gdf_gid = gdf[fnmatch.filter(cols, 'GID_*')].copy()
+    gdf_gid.dropna(axis=1, how='all', inplace=True)
+    # Last list element, Last character
+    last_populated_gid = gdf_gid.columns.values.tolist()[-1][-1]
+    print('\tMax Admin Level {0}'.format(last_populated_gid))
+    # Drop all cols after first empty GID_* column.
+    drop_cols = []
+    for col in cols:
+        if is_int(col[-1]) is True and int(col[-1]) > int(last_populated_gid):
+            drop_cols.append(col)
+    gdf.drop(drop_cols, axis=1, inplace=True)
+    # and take out ID_* fields
+    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'ID_*')], 
+             axis=1, inplace=True)
+    # and take out REMARKS_*
+#    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'REMARKS_*')], 
+#             axis=1, inplace=True)
+    # and take out VALID*
+#    gdf.drop(gdf[fnmatch.filter(gdf.columns.values.tolist(), 'VALID*')], 
+#             axis=1, inplace=True)
+    # and the UID
+    gdf.drop(columns=['UID'], axis=1, inplace=True)
+    # anything else is associated with the last avialable admin level if
+    # populated.
+    gdf.dropna(axis=1, how='all', inplace=True)
+    #
+    level = 0
+    while level < int(last_populated_gid):
+        print('\tProcessing Level {0}'.format(level))
+        cols_agg = []
+        cols_required = []
+        for col in gdf.columns.values.tolist():
+            if is_int(col[-1]) is True and int(col[-1]) <= level:
+                cols_agg.append(col)
+        # Include the geometry column 
+        cols_required = list(cols_agg)
+        cols_required.append('geometry')
+        # Dissolve using GID_* only - assume all other fields to use First
+        # value. Was dropping features where looking to be specific and
+        # groupby more fields (that should all be the same)
+        cols_agg = [x for x in cols_agg if 'GID_' in x]
+        gdf_dslv = gdf[cols_required].dissolve(by=cols_agg)
+        gdf_dslv.reset_index(inplace=True)
+        gdf_dslv = gdf_dslv[cols_required]
+        print('\tAggregated to {0} features.'.format(len(gdf_dslv)))
+        return_dict["a{0}".format(level)] = gdf_dslv 
+        # 
+        level += 1
+    # Include the highest level
+    return_dict["a{0}".format(last_populated_gid)] = gdf
+    return return_dict
+        
+    
+##
+# If gdf is passed only no way to know what the AOI country is.
+def get_neighbour_countries(gdf_source, gdf_A0=None, NAME_0=None, GID_0=None):
```

Comment:
`gdf_A0` shouldn't be set to `None` here since it's required

Reply:
Ok, I was thinking about extending this function so that it would be possible to pass a Name or GID depending on what was available to the function. Doing this would require the function to generate a gdf based on the Name or GID but I hadn't yet completed that option in the function.

Reply:
Ah okay, you can leave it in then, no worries.